### PR TITLE
Budget app

### PR DESCRIPTION
**Pull request by shahzaman900**
### Within this branch i have achieved the following:

- Developed a secure authentication system allowing users to log in with email and password, ensuring data privacy by restricting access to authorized users only.
- 
- Implemented a user-friendly category management feature, allowing users to add, delete, and customize categories with ease. Users can enrich their categories by adding names and image URLs, with helpful suggestions provided for seamless category creation.
- 
- Designed an intuitive navigation flow, guiding users from the category page to a dedicated transactions page. This enables users to add transactions specific to each category, fostering efficient monitoring of expenses.
- 
- Implemented a comprehensive spending tracking system, displaying the total amount for each category along with the overall spending, empowering users to manage their finances effectively.
- 
- Ensured the robustness of the project by conducting thorough testing using industry-standard tools such as Copybara for version control and RSpec for behavior-driven development, ensuring a reliable and bug-free application.
- 
- The design has been added using Gregory Vella provided template

### CODE REVIEW

File: app/controllers/purchases_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  def show
+    @each_purchase_nav = true
+    @category = Category.includes(:purchases).find(params[:category_id])
+    @purchase = @category.purchases.includes(:categories).find(params[:id])
+  end
+
+  def new
+    @new_purchase_nav = true
+    @category = Category.find(params[:category_id])
+    @purchase = purchase.new
```
Comment (tugsem) [inline app/controllers/purchases_controller.rb:17]: - [ ] Please fix this name error as it's not working.You are calling a class method and it should start with a capital letter.

Comment (tugsem) [inline app/controllers/purchases_controller.rb:17]: - [ ] Please fix this name error as it's not working.You are calling a class method and it should start with a capital letter.

File: README.md [hunk 2]
```diff
@@ -1,24 +1,162 @@
...
+
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+
+
+## 🚀 Live Demo & Video presentation<a name="live-demo"></a> <a name='video'></a>
+<br>
+
+- [Check the live demo, mobile version](https://budget-app-shah.onrender.com)
```
Comment (tugsem) [inline README.md:59]: - [ ] Please fix the live demo of your app as right now it's not working.

File: README.md [hunk 3]
```diff
@@ -1,24 +1,162 @@
...
+- [Check quick demonstration of the app video by clicking the link](https://www.loom.com/share/ccf39f75d66748c2b83c3fc12878fb83?sid=90f60902-56a8-410c-88fe-954c385f076aa)
```
Comment (tugsem) [inline README.md:60]: - [ ] Please add a new video link belongs to you as this link belongs to someone else.

File: README.md [hunk 4]
```diff
@@ -1,24 +1,162 @@
...
+
+<br>
+I would like to extend my heartfelt gratitude to Microverse for providing me with the opportunity to undertake this capstone project. Additionally, I am immensely thankful to the code reviewers who took the time to assess my work and provide valuable feedback, which greatly contributed to the refinement and success of the project. Your guidance and support have been invaluable throughout this solo endeavor, and I am truly grateful for your contributions.
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+<br>
+
+## 📝 License <a name="license"></a>
+
+This project is [MIT](./LICENSE) licensed.
```
Comment (tugsem) [inline README.md:162]: - [ ] Kindly fix the license link as it's not working.

**Review by tugsem — CHANGES_REQUESTED** [review]
##  STATUS: REQUIRED CHANGES ♻️
Hello @sheezu900 👋 , 

Good job so far! 👏 
There are some issues that you still need to work on to go to the next project but you are almost there!

Alt Text
## Highlights 🚀 
✔️linters are passing
✔️correct use of gitflow
✔️professional documentation
✔️ the design matches the design on behance

## Project documentation
❌ a description of the project.
❌ a demo of the project features.
❌ you should also highlight some interesting piece of code or something you built that you are very proud of.

## Required Changes ♻️
- [x] Please write unit and integration tests for all the most important components: 
image

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Felix45 — CHANGES_REQUESTED** [review]
Hi @sheezu900, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



 ## Highlights :heavy_check_mark:
 - [x] Application has a Good UI design :white_check_mark:
 - [x] Application functions as expected :white_check_mark:
- [x] Application uses `devise gem` for authentication :white_check_mark:
- [x] Pull request has a descriptive title and summary :white_check_mark:
- [x] Repository has a professional readme file :white_check_mark:
- [x] Implemented some of the changes from the previous reviewer :white_check_mark:

 ## Required Changes ♻️

- [Required] Good job making most of the required changes from the previous reviewer, however I've noticed that `tests` you added are not passing as expected. I've also noticed that the `rails-rspec` gem is missing from your gem file. This makes it difficult to run the tests using the `rspec spec` command. Please rectify the failing tests and ensure that they are passing when run using `rspec spec` command.

## Screenshot of failing tests

Screenshot from 2023-12-02 23-20-10

- [Required] Good job adding a video presentation to your project, however, you forgot to mention a piece of code that you are most proud of. According to the project requirements you are supposed to 

Screenshot from 2023-12-02 23-30-09




_Check the comments under the review._

## Optional suggestions

- [Optional] I think it would be nice if you add the `node_modules` directory to the `.gitignore` file and remove it from being tracked by `git`. Currently, this pull request has over `5000+` files which makes it difficult to post a review. I highly recommend you do this to make it easy for future reviewers to post their reviews.

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @sheezu900,




Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the changes, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- Great effort making the requested changes.✔️
- Nice design. ✔️
- Descriptive PR. ✔️


## Required Changes ♻️
_Check the comments below._

- [ ]  As requested by other reviewers, All your tests are failing, kindly fix them.
- [ ] Your video is missing the part you should show a piece of code you find interesting as requested by other viewers. please, kindly record another video to add this missing requirement.




## Optional suggestions

- N/A

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mwanawabangona — APPROVED** [review]
Hi @sheezu900 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:
Congratulations!  


## Highlights
*   No linter errors :heavy_check_mark:
*   Repo looks professional :briefcase:
*   Solid pull request scroll
*   Good quality commits
*   Good job implementing the requested changes!! :clap: :clap: :clap:

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**